File: backend/app/field_schema.py
"""
The fields of a cover page, derived from its markdown.

A port of `frontend/lib/cover-page-template.ts` and `frontend/lib/field-schema.ts`.
Both ends read the same markdown and must agree on every field name, because a
name only one side knows is a value the model fills in and the document never
receives. `test/field-schemas.json` at the repo root is asserted by both test
suites, so a change to either implementation that is not matched by the other
fails loudly rather than silently losing a term.

Pure: no FastAPI, no settings, no I/O beyond what the caller passes in.
"""

import re
from dataclasses import dataclass, field
from typing import Literal

HEADING = re.compile(r"^(#{1,3})\s+(.*)$")
LABEL = re.compile(r"^<label>(.*)</label>$")
OPTIONAL = re.compile(r"^<optional\s*/>$")
CHOICE = re.compile(r"^-\s+\[[ xX]\]\s+(.*)$")
SIGNING_STATEMENT = re.compile(r"^By signing this Cover Page.*$", re.MULTILINE)
ATTRIBUTION = re.compile(
    r"^Common Paper .*free to use under \[CC BY 4\.0\].*$", re.MULTILINE
)
PARTY_ROLES = re.compile(r"^\|\|\s*(.+?)\s*\|\s*(.+?)\s*\|?$")
PLACEHOLDER = re.compile(r"\[([^\]]*)\]")
LABELLED_LINE = re.compile(r"^(.+?):\s*\[(.*)\]\s*$")
GUIDANCE = re.compile(r"^(fill in|describe|list )", re.IGNORECASE)
DATE_PLACEHOLDER = re.compile(r"\bdates?\b", re.IGNORECASE)

MULTILINE_CHARS = 40
DEFAULT_PARTY_ROLES = ("Party 1", "Party 2")

PARTY_FIELDS = ("company", "name", "title", "notice_address")


@dataclass
class Section:
    heading: str
    hint: str | None = None
    choices: list[str] = field(default_factory=list)
    body: str = ""
    required: bool = True


@dataclass
class CoverPage:
    title: str
    intro: str
    sections: dict[str, Section]
    party_roles: tuple[str, str]


@dataclass
class ChoiceOption:
    sentence: str
    unit: str | None


@dataclass
class GroupLine:
    key: str
    label: str
    placeholder: str


@dataclass
class FieldSpec:
    kind: Literal["text", "date", "choice", "group"]
    key: str
    label: str
    hint: str | None
    required: bool
    initial: str = ""
    placeholder: str = ""
    multiline: bool = False
    options: list[ChoiceOption] = field(default_factory=list)
    lines: list[GroupLine] = field(default_factory=list)

    def wire_keys(self) -> list[str]:
        """Every name this field goes by on the wire."""
        if self.kind == "choice":
            return [f"{self.key}_option", f"{self.key}_number"]
        if self.kind == "group":
            return [line.key for line in self.lines]
        return [self.key]


@dataclass
class DocumentSchema:
    title: str
    party_roles: tuple[str, str]
    fields: list[FieldSpec]

    def by_key(self) -> dict[str, FieldSpec]:
        return {key: spec for spec in self.fields for key in spec.wire_keys()}


def slugify(heading: str) -> str:
    """`"Governing Law & Jurisdiction"` becomes `"governing_law_jurisdiction"`."""
    return re.sub(r"[^a-z0-9]+", "_", heading.lower()).strip("_")


def placeholder_of(text: str) -> str | None:
    match = PLACEHOLDER.search(text)
    return match.group(1) if match else None


def placeholder_unit(text: str) -> str:
    return re.sub(r"^\s*\d+\s*", "", placeholder_of(text) or "")


def parse_cover_page(markdown: str) -> CoverPage:
    signing = SIGNING_STATEMENT.search(markdown)
    attribution = ATTRIBUTION.search(markdown)
    signing_text = signing.group(0) if signing else ""
    attribution_text = attribution.group(0) if attribution else ""

    title = ""
    intro_lines: list[str] = []
    sections: dict[str, Section] = {}
    current: Section | None = None
    party_roles: tuple[str, str] | None = None
    body_lines: list[str] = []

    def flush() -> None:
        if current is not None:
            # Sections are keyed by heading, so a repeat would overwrite the
            # first and quietly drop a field somebody could have filled in.
            if current.heading in sections:
                raise SchemaError(f'Two sections both headed "{current.heading}"')
            current.body = "\n".join(body_lines).strip()
            sections[current.heading] = current
            body_lines.clear()

    for line in markdown.split("\n"):
        heading = HEADING.match(line)
        if heading:
            hashes, text = heading.group(1), heading.group(2)
            if hashes == "#":
                title = text
            elif hashes == "###":
                flush()
                current = Section(heading=text)
            continue

        if (
            line.startswith("|")
            or (signing_text and line == signing_text)
            or (attribution_text and line == attribution_text)
        ):
            roles = PARTY_ROLES.match(line)
            if roles and party_roles is None:
                party_roles = (roles.group(1), roles.group(2))
            continue

        if current is None:
            intro_lines.append(line)
            continue

        label = LABEL.match(line.strip())
        if label:
            current.hint = label.group(1)
            continue

        if OPTIONAL.match(line.strip()):
            current.required = False
            continue

        choice = CHOICE.match(line.strip())
        if choice:
            current.choices.append(choice.group(1).strip())
            continue

        body_lines.append(line)

    flush()

    return CoverPage(
        title=title,
        intro="\n".join(intro_lines).strip(),
        sections=sections,
        party_roles=party_roles or DEFAULT_PARTY_ROLES,
    )


def _labelled_lines(body: str) -> list[tuple[str, str]]:
    matches = (LABELLED_LINE.match(line.strip()) for line in body.split("\n"))
    return [(m.group(1), m.group(2)) for m in matches if m]


def _hint_for(section: Section) -> str | None:
    if section.hint:
        return section.hint
    if section.body and "[" not in section.body:
        return section.body
    return None


def _classify(section: Section) -> FieldSpec:
    common = {
        "key": slugify(section.heading),
        "label": section.heading,
        "hint": _hint_for(section),
        "required": section.required,
    }

    if len(section.choices) >= 2:
        return FieldSpec(
            kind="choice",
            options=[
                ChoiceOption(
                    sentence=sentence,
                    unit=(
                        placeholder_unit(sentence)
                        if any(c.isdigit() for c in (placeholder_of(sentence) or ""))
                        else None
                    ),
                )
                for sentence in section.choices
            ],
            **common,
        )

    lines = _labelled_lines(section.body)
    if len(lines) >= 2:
        return FieldSpec(
            kind="group",
            lines=[
                GroupLine(key=slugify(label), label=label, placeholder=placeholder)
                for label, placeholder in lines
            ],
            **common,
        )

    placeholder = placeholder_of(section.body) or ""
    if placeholder and DATE_PLACEHOLDER.search(placeholder):
        return FieldSpec(kind="date", **common)

    guidance = bool(GUIDANCE.match(placeholder))
    return FieldSpec(
        kind="text",
        initial="" if guidance else placeholder,
        placeholder=placeholder if guidance else "",
        multiline=not placeholder or len(placeholder) > MULTILINE_CHARS,
        **common,
    )


class SchemaError(Exception):
    """A cover page that cannot be turned into a usable set of fields."""


def parse_document_schema(markdown: str) -> DocumentSchema:
    cover_page = parse_cover_page(markdown)
    fields = [_classify(section) for section in cover_page.sections.values()]

    seen: set[str] = set()
    for spec in fields:
        for key in spec.wire_keys():
            if key in seen:
                raise SchemaError(f'{cover_page.title}: two fields both named "{key}".')
            seen.add(key)

    return DocumentSchema(
        title=cover_page.title,
        party_roles=cover_page.party_roles,
        fields=fields,
    )

File: frontend/lib/field-schema.ts
/**
 * Turns a parsed cover page into the typed fields the app fills in.
 *
 * `cover-page-template` reads the markdown's grammar — headings, hints,
 * choices, brackets. This module decides what each section *means*: whether it
 * is a date, a pick between the template's own sentences, a pair of labelled
 * lines, or free text. Everything downstream — the form, the chat, the preview,
 * the PDF — reads these specs rather than naming any document's fields, which
 * is what lets one set of components serve all eleven agreements.
 *
 * The dialect the classification relies on, all of it already Common Paper's:
 *
 * - `- [ ]` alternatives                       -> a choice between them
 * - `Label: [guidance]` lines, two or more     -> a group of labelled lines
 * - a lone `[placeholder]` mentioning a date   -> a date
 * - anything else                              -> free text
 *
 * A bracket is a suggested value the user can keep, except when it opens with
 * "Fill in", which marks it as guidance about what to write instead. That
 * distinction is Common Paper's own: the Mutual NDA offers
 * `[Evaluating whether to enter into a business relationship...]` as a real
 * default but `[Fill in state]` as an instruction.
 *
 * Pure and dependency-free, and mirrored by `backend/app/field_schema.py`. The
 * two are held together by `test/field-schemas.json`, which both test suites
 * assert against, so neither can drift without a red test.
 */

import {
  type CoverPageSection,
  type CoverPageTemplate,
  placeholderOf,
  placeholderUnit,
} from "./cover-page-template";

export interface ChoiceOption {
  /** The template's own sentence, with any `[placeholder]` left intact. */
  sentence: string;
  /** The unit after the number, e.g. `"year(s)"`, when the sentence takes one. */
  unit: string | null;
}

/** One `Label: [guidance]` line within a grouped field. */
export interface GroupLine {
  key: string;
  label: string;
  /** The bracketed guidance, shown in the input rather than entered into it. */
  placeholder: string;
}

interface FieldCommon {
  /** Wire name, from the heading. Stable across frontend and backend. */
  key: string;
  /** The heading as written, used as the caption and in the document. */
  label: string;
  hint: string | null;
  required: boolean;
}

export type FieldSpec =
  | (FieldCommon & {
      kind: "text";
      /** Suggested wording the user may keep, or `""`. */
      initial: string;
      /** Guidance shown in an empty input. */
      placeholder: string;
      multiline: boolean;
    })
  | (FieldCommon & { kind: "date" })
  | (FieldCommon & { kind: "choice"; options: ChoiceOption[] })
  | (FieldCommon & { kind: "group"; lines: GroupLine[] });

export interface DocumentSchema {
  title: string;
  intro: string;
  signingStatement: string;
  attribution: string;
  partyRoles: [string, string];
  fields: FieldSpec[];
}

/**
 * A heading as a wire name: `"Governing Law & Jurisdiction"` becomes
 * `"governing_law_jurisdiction"`.
 *
 * snake_case rather than kebab because these become attribute names on the
 * backend's generated Pydantic model, where a hyphen is not a legal identifier.
 */
export function slugify(heading: string): string {
  return heading
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "_")
    .replace(/^_+|_+$/g, "");
}

/** A `Label: [guidance]` line. */
const LABELLED_LINE = /^(.+?):\s*\[(.*)\]\s*$/;
/** Opens a bracket that instructs rather than suggests. */
const GUIDANCE = /^(fill in|describe|list )/i;
/**
 * A bracket asking for a date.
 *
 * Anchored to whole words: an unanchored `/date/` also matches "updates",
 * which would turn a text field into a date picker.
 */
const DATE_PLACEHOLDER = /\bdates?\b/i;
/** Beyond this, a suggested value reads as prose and wants room to edit. */
const MULTILINE_CHARS = 40;

function labelledLines(body: string): { label: string; placeholder: string }[] {
  return body
    .split("\n")
    .map((line) => LABELLED_LINE.exec(line.trim()))
    .filter((match): match is RegExpExecArray => match !== null)
    .map((match) => ({ label: match[1], placeholder: match[2] }));
}

function hintFor(section: CoverPageSection): string | null {
  if (section.hint) return section.hint;
  // Prose with nothing to fill in is guidance about the field, not part of it
  // — "List any modifications to the MNDA" reads as help text.
  if (section.body && !section.body.includes("[")) return section.body;
  return null;
}

function classify(section: CoverPageSection): FieldSpec {
  const key = slugify(section.heading);
  const common = {
    key,
    label: section.heading,
    hint: hintFor(section),
    required: section.required,
  };

  if (section.choices.length >= 2) {
    return {
      ...common,
      kind: "choice",
      options: section.choices.map((sentence) => ({
        sentence,
        // Only a bracket holding a number is something the user fills in; the
        // rest of the sentence is the template's wording and stays as written.
        unit: /\d/.test(placeholderOf(sentence) ?? "")
          ? placeholderUnit(sentence)
          : null,
      })),
    };
  }

  const lines = labelledLines(section.body);
  if (lines.length >= 2) {
    return {
      ...common,
      kind: "group",
      lines: lines.map((line) => ({
        key: slugify(line.label),
        label: line.label,
        placeholder: line.placeholder,
      })),
    };
  }

  const placeholder = placeholderOf(section.body) ?? "";
  if (placeholder && DATE_PLACEHOLDER.test(placeholder)) {
    return { ...common, kind: "date" };
  }

  const guidance = GUIDANCE.test(placeholder);
  return {
    ...common,
    kind: "text",
    initial: guidance ? "" : placeholder,
    placeholder: guidance ? placeholder : "",
    multiline: !placeholder || placeholder.length > MULTILINE_CHARS,
  };
}

/** Raised when a cover page cannot be turned into a usable set of fields. */
export class SchemaError extends Error {}

/**
 * The document's fields, in the order the cover page lists them.
 *
 * Throws rather than dropping anything: a cover page whose fields collide would
 * otherwise lose a value silently, and the place to find that out is the build,
 * not a signed agreement with a gap in it.
 */
export function parseDocumentSchema(
  template: CoverPageTemplate,
): DocumentSchema {
  const fields = Object.values(template.sections).map(classify);

  const seen = new Set<string>();
  for (const key of fields.flatMap(fieldKeys)) {
    if (seen.has(key)) {
      throw new SchemaError(
        `${template.title}: two fields both named "${key}". Rename one heading.`,
      );
    }
    seen.add(key);
  }

  return {
    title: template.title,
    intro: template.intro,
    signingStatement: template.signingStatement,
    attribution: template.attribution,
    partyRoles: template.partyRoles,
    fields,
  };
}

/** Wire names for a choice field: which sentence, and the number inside it. */
export const optionKey = (key: string) => `${key}_option`;
export const numberKey = (key: string) => `${key}_number`;

/**
 * Every wire name a field owns.
 *
 * One name for a text or date field, one per line for a group, and two for a
 * choice — which sentence, and the number that goes in it. The backend derives
 * exactly these, which is what the shared fixture pins.
 */
export function fieldKeys(field: FieldSpec): string[] {
  switch (field.kind) {
    case "choice":
      return [optionKey(field.key), numberKey(field.key)];
    case "group":
      return field.lines.map((line) => line.key);
    default:
      return [field.key];
  }
}
